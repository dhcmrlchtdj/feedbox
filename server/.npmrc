disturl=https://npm.taobao.org/mirrors/node
fse_binary_host_mirror=https://npm.taobao.org/mirrors/fsevents
node_sqlite3_binary_host_mirror=https://npm.taobao.org/mirrors
