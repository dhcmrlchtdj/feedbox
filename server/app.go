package server

import (
	"context"
	"encoding/json"
	"os"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/expvar"
	"github.com/gofiber/fiber/v2/middleware/pprof"
	"github.com/gofiber/fiber/v2/middleware/recover"
	"github.com/gofiber/fiber/v2/middleware/requestid"
	"github.com/pkg/errors"
	"github.com/rs/zerolog"

	"github.com/dhcmrlchtdj/feedbox/internal/database"
	"github.com/dhcmrlchtdj/feedbox/internal/global"
	"github.com/dhcmrlchtdj/feedbox/server/handler"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/auth/cookie"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/auth/github"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/etag"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/logger"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/secure"
	"github.com/dhcmrlchtdj/feedbox/server/middleware/validate"
	"github.com/dhcmrlchtdj/feedbox/server/types"
)

func Create(ctx context.Context) *fiber.App {
	appConfig := fiber.Config{
		DisableStartupMessage: true,
		// BodyLimit: 4 * 1024 * 1024,
		// Concurrency: 256 * 1024,
		ErrorHandler:  errorHandler,
		StrictRouting: true,
		CaseSensitive: true,
		// ProxyHeader: "X-Forwarded-For"
	}
	app := fiber.New(appConfig)

	// middleware
	app.Use(recover.New())
	app.Use(requestid.New())
	app.Use(logger.New(ctx))
	app.Use(etag.New())
	app.Use(secure.New())
	if os.Getenv("ENV") != "prod" {
		app.Use(pprof.New())
		app.Use(expvar.New())
	}

	// router
	subDir := os.Getenv("SERVER_SUB_DIR")
	if subDir == "" {
		subDir = "/"
	}
	appRouter := app.Group(subDir)
	setupRoute(appRouter) // nolint:contextcheck

	return app
}

func setupRoute(app fiber.Router) {
	// API
	app.Use("/api/v1", cookie.New(cookie.Config{
		Name:      "token",
		Validator: cookieValidator,
	}))
	app.Get("/api/v1/user", handler.UserInfo)
	app.Get("/api/v1/feeds", handler.FeedList)
	app.Put("/api/v1/feeds/add", validate.ContentType("application/json"), handler.FeedAdd)
	app.Delete("/api/v1/feeds/remove", validate.ContentType("application/json"), handler.FeedRemove)
	app.Get("/api/v1/feeds/export", handler.FeedExport)
	app.Post("/api/v1/feeds/import", validate.ContentType("multipart/form-data"), handler.FeedImport)

	// Auth
	app.Get("/api/logout", handler.Logout)
	app.Get(
		"/api/connect/github",
		github.New(github.Config{
			ClientID:     os.Getenv("GITHUB_CLIENT_ID"),
			ClientSecret: os.Getenv("GITHUB_CLIENT_SECRET"),
		}),
		handler.ConnectGithub,
	)

	// debug
	if os.Getenv("ENV") != "prod" {
		app.Get("/helper/preview", handler.HelperParseFeed)
	}

	// telegram
	app.Post(
		"/webhook/telegram/"+os.Getenv("TELEGRAM_WEBHOOK_PATH"),
		validate.ContentType("application/json"),
		handler.TelegramWebhook,
	)

	// static
	app.Get("/", handler.StaticFile("_build/index.html", handler.StaticWithCustomHeader("_build/index.html.json")))
	app.Get("/sw.js", handler.StaticFile("_build/sw.js"))
	app.Get("/sw.js.map", handler.StaticFile("_build/sw.js.map"))
	app.Get("/favicon.ico", handler.StaticFile("_build/favicon.svg", handler.StaticWithMaxAge(60*60*24*7)))
	app.Get("/:filename", handler.StaticDir("_build/", handler.StaticWithMaxAge(60*60*24*7)))
}

func errorHandler(c *fiber.Ctx, err error) error {
	notCare := fiber.DefaultErrorHandler(c, err)
	code := c.Response().StatusCode()
	if code >= 500 {
		zerolog.Ctx(c.UserContext()).Error().Str("module", "app").Stack().Err(err).Send()
	}
	return notCare
}

func cookieValidator(ctx context.Context, tokenStr string) ( /* Credential */ any, error) {
	plaintext, err := global.Sign.DecodeFromHex(tokenStr)
	if err != nil {
		return nil, errors.New("invalid token")
	}
	var credential types.Credential
	if err := json.Unmarshal(plaintext, &credential); err != nil {
		return nil, errors.New("broken token")
	}
	if time.Now().Unix() > credential.ExpiresAt {
		return nil, errors.New("expired token")
	}

	if _, err = global.Database.GetUserByID(ctx, credential.UserID); err != nil {
		if errors.Is(err, database.ErrEmptyRow) {
			err = errors.New("invalid user")
		}
		return nil, err
	}

	return credential, nil
}
