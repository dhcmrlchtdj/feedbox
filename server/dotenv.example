CRON_USERNAME="username"
CRON_PASSWORD="crypto.randomBytes(2).toString('hex')"

SITE="http://localhost:9000"
API="http://localhost:8000"

COOKIE_SECRET="crypto.randomBytes(2).toString('hex')"

GITHUB_AUTH_SECRET="crypto.randomBytes(2).toString('hex')"
GITHUB_CLIENT_ID="https://github.com/settings/developers"
GITHUB_CLIENT_SECRET="https://github.com/settings/developers"

MAILGUN_API_KEY="key-example"
MAILGUN_DOMAIN="feedbox.example.com"
MAILGUN_FROM="feedbox <feedbox@example.com>"

DATABASE_URL="postgres://example"
