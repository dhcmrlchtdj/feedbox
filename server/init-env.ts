import * as path from 'path'
import * as dotenv from 'dotenv-safe'

dotenv.config({
    // '*.ts' will be compiled to '_build/*.js'
    path: path.resolve(__dirname, '../../dotenv'),
    example: path.resolve(__dirname, '../../dotenv.example'),
})
