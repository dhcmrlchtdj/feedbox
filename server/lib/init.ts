import 'reflect-metadata'
import * as path from 'path'
import * as dotenv from 'dotenv-safe'
import initDB from './models'

export default async () => {
    if (process.env.NODE_ENV !== 'production') {
        dotenv.config({
            path: path.resolve(__dirname, '../dotenv'),
            example: path.resolve(__dirname, '../dotenv.example'),
        })
    }

    await initDB()
}

process.on('unhandledRejection', err => {
    console.error(err)
    process.exit(1)
})
