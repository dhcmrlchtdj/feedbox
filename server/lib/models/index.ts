import conn from './conn'

interface User {
    id: number
    github_id: number
    email: string
}

interface Feed {
    id: number
    url: string
    latest_checked: Date
    latest_updated: Date
}

export interface Link {
    id: number
    url: string
    feed_id: number
}

interface RUserFeed {
    user_id: number
    feed_id: number
}

const db = {
    async init() {
        const k = await conn()
        await k.schema
            .dropTableIfExists('RUserFeed')
            .dropTableIfExists('Link')
            .dropTableIfExists('Feed')
            .dropTableIfExists('User')
            .createTable('User', table => {
                table.increments('id')
                table.dateTime('created_at').defaultTo(k.fn.now())
                table
                    .integer('github_id')
                    .notNullable()
                    .unique()
                table
                    .string('email', 255)
                    .notNullable()
                    .unique()
            })
            .createTable('Feed', table => {
                table.increments('id')
                table.dateTime('created_at').defaultTo(k.fn.now())
                table
                    .string('url', 255)
                    .notNullable()
                    .unique()
                table
                    .dateTime('latest_checked')
                    .nullable()
                    .defaultTo(null)
                table
                    .dateTime('latest_updated')
                    .nullable()
                    .defaultTo(null)
            })
            .createTable('Link', table => {
                table.increments('id')
                table.dateTime('created_at').defaultTo(k.fn.now())
                table.integer('url', 255).notNullable()
                table.integer('feed_id').unsigned()
                table.foreign('feed_id').references('Feed.id')
            })
            .createTable('RUserFeed', table => {
                table.increments('id')
                table.dateTime('created_at').defaultTo(k.fn.now())
                table.integer('user_id').unsigned()
                table.integer('feed_id').unsigned()
                table.unique(['user_id', 'feed_id'])
                table.foreign('user_id').references('User.id')
                table.foreign('feed_id').references('Feed.id')
            })
    },

    async getUserById(id: number): Promise<User | null> {
        const r = await conn()
            .select('id', 'github_id as githubId', 'email')
            .from('User')
            .where({ id })
        if (r.length > 0) {
            return r[0]
        } else {
            return null
        }
    },

    async getUserIdByGithub(github_id: number, email: string): Promise<number> {
        await conn().raw(
            `INSERT INTO User(github_id,email) VALUES(?,?)
            ON CONFLICT(github_id) DO UPDATE SET email=?`,
            [github_id, email, email],
        )
        const r = await conn()
            .select('id')
            .from('User')
            .where({ email })
        return r[0].id
    },

    async getFeedIdByUrl(url: string): Promise<number> {
        await conn().raw(
            `INSERT INTO Feed(url) VALUES(?)
            ON CONFLICT DO NOTHING`,
            [url],
        )
        const r = await conn()
            .select('id')
            .from('Feed')
            .where({ url })
        return r[0].id
    },

    async getFeedByUser(userId: number): Promise<Feed[]> {
        const r = await conn()
            .select(
                'Feed.id as id',
                'Feed.url as url',
                'Feed.latest_checked as latest_checked',
                'Feed.latest_updated as latest_updated',
            )
            .from('RUserFeed')
            .innerJoin('User', 'User.id', 'RUserFeed.user_id')
            .innerJoin('Feed', 'Feed.id', 'RUserFeed.feed_id')
            .where('User.id', userId)
        return r
    },

    async getFeedForUpdate(): Promise<Map<Feed, User[]>> {
        // TODO
        const r = await conn()
            .select('feed_id', 'user_id')
            .from<RUserFeed>('r_user_feed')
        const feeds = [] as Feed[]
        const feedMap = feeds.reduce((map, feed) => {
            map[feed.id] = feed
            return map
        }, new Map<number, Feed>())
        const users = [] as User[]
        const userMap = users.reduce((map, user) => {
            map[user.id] = user
            return map
        }, new Map<number, User>())
        const o = r.reduce((map, { feedId, userId }) => {
            const feed = feedMap.get(feedId)!!
            const user = userMap.get(userId)!!
            const arr = map.get(feed) ?? []
            arr.push(user)
            map.set(feed, arr)
            return map
        }, new Map<Feed, User[]>())
        return o
    },

    async subscribe(user_id: number, feed_id: number) {
        await conn().raw(
            `INSERT INTO RUserFeed(user_id, feed_id) VALUES(?,?)
            ON CONFLICT DO NOTHING`,
            [user_id, feed_id],
        )
    },

    async unsubscribe(user_id: number, feed_id: number) {
        await conn()
            .from('RUserFeed')
            .where({ user_id, feed_id })
            .del()
    },

    async subscribeUrls(user_id: number, urls: string[]) {
        const qvalue = urls.map(_ => '(?)').join(',')
        await conn().raw(
            `INSERT INTO Feed(url) VALUES ${qvalue}
            ON CONFLICT DO NOTHING`,
            urls,
        )

        const qrange = urls.map(_ => '?').join(',')
        await conn().raw(
            `INSERT INTO RUserFeed(user_id, feed_id)
            SELECT ?, Feed.id FROM Feed WHERE Feed.url in (${qrange})
            ON CONFLICT DO NOTHING`,
            [user_id, ...urls],
        )
    },
}

export default db
