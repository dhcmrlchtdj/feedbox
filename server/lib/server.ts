import * as Hapi from '@hapi/hapi'
import plugins from './plugins'
import routes from './routes'
import prepare from './prepare'
import Model from './models'

const common = async () => {
    await prepare()

    const corsConf = {
        headers: [
            'accept',
            'authorization',
            'content-type',
            'if-none-match',
            'content-type',
            'x-sw-strategy',
            'x-sw-race',
            'x-sw-actions',
        ],
        credentials: true,
    }
    const server = Hapi.server({
        port: Number(process.env.PORT || 8000),
        host: '0.0.0.0',
        routes: {
            payload: { allow: 'application/json' },
            cors: process.env.SERVER === process.env.WEB ? false : corsConf,
        },
    })
    await plugins(server)
    server.route(routes)

    server.events.on('stop', async () => {
        await Model.destroy()
    })

    return server
}

export const init = async () => {
    const server = await common()
    server.initialize()
    return server
}

export const start = async () => {
    const server = await common()
    server.start()
    return server
}
