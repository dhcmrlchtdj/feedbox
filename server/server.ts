import './init-env'
import * as path from 'path'
import * as Hapi from '@hapi/hapi'
import { addPlugins } from './plugins'
import { routes } from './routes'
import { model } from './models'
import { telegramBot } from './telegram/bot'

const common = async () => {
    const server = Hapi.server({
        port: Number(process.env.PORT || 8000),
        host: '0.0.0.0',
        routes: {
            payload: { allow: 'application/json' },
            files: {
                relativeTo: path.resolve(__dirname, '../static'),
            },
        },
    })
    await addPlugins(server)
    server.route(routes)

    server.events.on('stop', async () => {
        await model.destroy()
    })

    return server
}

export const init = async () => {
    const server = await common()
    server.initialize()
    return server
}

export const start = async () => {
    const server = await common()
    server.start()

    if (process.env.NODE_ENV === 'production') {
        // init telegram webhook
        await telegramBot.registerWebhook()
    }

    return server
}
