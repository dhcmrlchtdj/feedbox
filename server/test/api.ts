import { init } from '../lib/server'
import User from '../lib/models/user'

let server: any
let userId: number
beforeAll(async () => {
    server = await init()
    const user = await User.takeOrCreateByGithub(1, 'user@example.com')
    userId = user.id
})
afterAll(async () => {
    await server.stop()
})

describe('user API', () => {
    test('/api/v1/user', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/user',
            auth: {
                strategy: 'session',
                credentials: { userId },
            },
        })
        expect(resp.result).toMatchInlineSnapshot(`
            User {
              "email": "user@example.com",
              "githubId": 1,
              "id": 1,
            }
        `)
    })

    test('/api/logout', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/logout',
            headers: {
                Cookie: 'token=***',
            },
        })
        expect(resp.headers['set-cookie']).toMatchInlineSnapshot(`
            Array [
              "token=; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT; HttpOnly; Path=/api",
            ]
        `)
        expect(resp.headers['location']).toMatchInlineSnapshot(
            `"http://localhost:9000"`,
        )
        expect(resp.statusCode).toMatchInlineSnapshot(`302`)
    })
})

describe('feed API', () => {
    test('/api/v1/feeds/add', async () => {
        const resp = await server.inject({
            method: 'put',
            url: '/api/v1/feeds/add',
            payload: {
                url: 'https://example.com/rss',
            },
            auth: {
                strategy: 'session',
                credentials: { userId },
            },
        })
        expect(resp.result).toMatchInlineSnapshot(`
            Array [
              Feed {
                "id": 1,
                "lastCheck": null,
                "lastUpdated": null,
                "url": "https://example.com/rss",
              },
            ]
        `)
    })

    test('/api/v1/feeds', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/feeds',
            auth: {
                strategy: 'session',
                credentials: { userId },
            },
        })
        expect(resp.result).toMatchInlineSnapshot(`
            Array [
              Feed {
                "id": 1,
                "lastCheck": null,
                "lastUpdated": null,
                "url": "https://example.com/rss",
              },
            ]
        `)
    })

    test('/api/v1/feeds/export', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/feeds/export',
            auth: {
                strategy: 'session',
                credentials: { userId },
            },
        })
        expect(resp.result).toMatchInlineSnapshot(`
            "<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
            <opml version=\\"1.0\\">
            <head><title>feeds</title></head>
            <body>
            <outline type=\\"rss\\" text=\\"example.com\\" xmlUrl=\\"https://example.com/rss\\"/>
            </body>
            </opml>"
        `)
    })

    test('/api/v1/feeds/remove', async () => {
        const resp = await server.inject({
            method: 'delete',
            url: '/api/v1/feeds/remove',
            payload: {
                feedId: 1,
            },
            auth: {
                strategy: 'session',
                credentials: { userId },
            },
        })
        expect(resp.result).toMatchInlineSnapshot(`Array []`)
    })
})
