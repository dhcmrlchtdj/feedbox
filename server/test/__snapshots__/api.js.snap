// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feed API /api/v1/feeds 1`] = `
Array [
  Object {
    "id": 1,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss",
  },
]
`;

exports[`feed API /api/v1/feeds/add 1`] = `
Array [
  Object {
    "id": 1,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss",
  },
]
`;

exports[`feed API /api/v1/feeds/export 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<opml version=\\"1.0\\">
<head><title>feeds</title></head>
<body>
<outline type=\\"rss\\" text=\\"example.com\\" xmlUrl=\\"https://example.com/rss\\"/>
</body>
</opml>"
`;

exports[`feed API /api/v1/feeds/import 1`] = `
Array [
  Object {
    "id": 1,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss",
  },
  Object {
    "id": 2,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`feed API /api/v1/feeds/remove 1`] = `
Array [
  Object {
    "id": 2,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "latest_checked": null,
    "latest_updated": null,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`user API /api/logout 1`] = `
Array [
  "token=; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT; HttpOnly; Path=/api",
]
`;

exports[`user API /api/logout 2`] = `"http://localhost:9000"`;

exports[`user API /api/logout 3`] = `302`;

exports[`user API /api/v1/user 1`] = `
Object {
  "email": "user@example.com",
  "githubId": 1,
  "id": 1,
}
`;
